[DX-285] React Native push notifications guide (#3228)
ably/docs: src/pages/docs/push/getting-started/react-native.mdx

</Aside>

### Handle push notifications <a id="step-2-handle"/>

The FCM SDK handles background push notifications automatically and displays them as system notifications. For foreground handling, use `@notifee/react-native` to display notifications while the app is open.

Add the following inside `PushScreen`:

<Code fixed="true">
```react
useEffect(() => {
  // Create a default Android notification channel
  if (Platform.OS === 'android') {
    notifee.createChannel({id: 'default', name: 'Default Channel'});
  }

  // Handle foreground push messages
  const unsubscribe = messaging().onMessage(async remoteMessage => {
    const title = remoteMessage.notification?.title ?? 'Push Notification';
    const body = remoteMessage.notification?.body ?? '';
    addLog(`Push received: ${title} — ${body}`);
    await notifee.displayNotification({
      title,
      body,
      android: {channelId: 'default'},
    });
  });

  return () => {
    unsubscribe();
  };
}, []);
```
</Code>

## Step 3: Subscribe and test push notifications <a id="step-3"/>

> ### Subscribe to channel push notifications <a id="step-3-subscribe"/>

Reviewer: ```suggestion
## Subscribe to channel push notifications <a id="step-3"/>
```

Author: applied

To subscribe your device to a channel so it can receive channel-based push notifications, add the following functions inside `PushScreen`, after the functions from Step 2:

> <Code fixed="true">

Reviewer: You only need `fixed=true` for things like `text` - not for every code block.

Author: removed for `react` code blocks

```react
async function subscribeToChannel() {
  try {
    await client.push.admin.channelSubscriptions.save({
      deviceId,
      channel: CHANNEL_NAME,
    });
    showStatus(`Subscribed to push on channel: ${CHANNEL_NAME}`);
  } catch (error) {
    showStatus(`Failed to subscribe: ${(error as Ably.ErrorInfo).message}`);
  }
}

async function unsubscribeFromChannel() {
  try {
    await client.push.admin.channelSubscriptions.remove({
      deviceId,
      channel: CHANNEL_NAME,
    });
    showStatus(`Unsubscribed from push on channel: ${CHANNEL_NAME}`);
  } catch (error) {
    showStatus(`Failed to unsubscribe: ${(error as Ably.ErrorInfo).message}`);
  }
}
```
</Code>

### Build the UI <a id="step-3-ui"/>

> Update the `return` statement in `PushScreen` to add buttons that call all the push functions:

Reviewer: ```suggestion
### Build the UI <a id="step-3-ui"/>

Build a UI in your app to add buttons that call all push functions. Update the `return` statement in `PushScreen`:
```

Do you think this should just be a step in its own right? That would make this step 4 and then publish step 5?

Author: done (back to 5 steps)

        </TouchableOpacity>
      </View>
      <ScrollView style={styles.logBox}>
        {log.map((entry, i) => (
          <Text key={i} style={styles.logEntry}>{entry}</Text>
        ))}
      </ScrollView>
      <TouchableOpacity style={[styles.btn, styles.btnGray]} onPress={() => setLog([])}>
        <Text style={styles.btnText}>Clear Log</Text>
      </TouchableOpacity>
    </View>
  </SafeAreaView>
);
```
</Code>

Add the button styles to the `StyleSheet.create` call:

<Code fixed="true">
```react
buttons: {gap: 8, marginBottom: 12},
btn: {padding: 14, borderRadius: 6, alignItems: 'center'},
btnText: {color: '#fff', fontWeight: '600'},
btnGreen: {backgroundColor: '#28a745'},
btnRed: {backgroundColor: '#dc3545'},
btnPurple: {backgroundColor: '#6f42c1'},
btnOrange: {backgroundColor: '#fd7e14'},
btnGray: {backgroundColor: '#6c757d', marginTop: 8},

```
</Code>

Build and run your app on a physical device:

<Code fixed="true">
```shell
# Android
npx react-native run-android

# iOS
npx react-native run-ios --device
```
</Code>

You can also open each platform project in their respective IDEs and run from there.

> ### Publish a push notification <a id="step-3-publish"/>

Reviewer: ```suggestion
## Step 4: Publish a push notification <a id="step-4"/>
```

Author: done

In the app tap **Activate Push** and wait until the status message displays your device ID.

> #### Direct push notification <a id="step-3-direct"/>

Reviewer: ```suggestion
### Publish directly to your device <a id="step-4-direct"/>
```

Author: done

Publish a push notification directly to your client ID (or device ID using `--device-id` instead of `--client-id`) via the Ably CLI:

<Code fixed="true">
```shell
ably push publish --client-id push-tutorial-client \
  --title "Hello" \
  --body "World!" \
  --data '{"foo":"bar"}'
```
</Code>

> #### Channel push notification <a id="step-3-channel"/>

Reviewer: ```suggestion
### Publish via a channel <a id="step-4-channel"/>
```

Author: done

---
title: "Getting started: Push Notifications in React Native"
meta_description: "Get started with Ably Push Notifications in React Native. Learn how to register for push notifications with Firebase Cloud Messaging (FCM), activate push on your client, handle incoming notifications, and send push messages on iOS and Android."
meta_keywords: "Push Notifications React Native, Ably Push, FCM, Android Push, iOS Push, React Native push notifications, Firebase Cloud Messaging, Ably Push Notifications guide, realtime push React Native, push notification example, Ably tutorial React Native, device registration, push messaging"
---

This guide will get you started with Ably Push Notifications in a new React Native application.

You'll learn how to set up your application with Firebase Cloud Messaging (FCM), register devices with Ably, send push notifications, subscribe to channel-based push, and handle incoming notifications on both iOS and Android.

## Prerequisites <a id="prerequisites"/>

1. [Sign up](https://ably.com/signup) for an Ably account.
2. Create a [new app](https://ably.com/accounts/any/apps/new), and create your first API key in the **API Keys** tab of the dashboard.

> 3. Your API key needs the `publish`, `subscribe`. Also add `push-admin` capabilities to test sending push notifications (in production, use your backend server to send push notifications by publishing messages with an `extras.push` field to a channel).

Reviewer: ```suggestion
  * Your API key needs the `publish`, `subscribe` capabilities. 
  * Also add the `push-admin` capability if you're using the same API key to send a push notification. In production this would more likely be a server using a different API key.
```

Author: done

> 4. For channel-based push, add a rule for the channel with **Push notifications enabled** checked. In the dashboard left sidebar: **Configuration** → **Rules** → **Add** or **Edit** a rule, then enable the Push notifications option. See [channel rules](/docs/channels#rules) for details.

Reviewer: ```suggestion
3. Add a rule to a channel so you can test sending push notification via a channel. Select [**Rules**](https://ably.com/accounts/any/apps/any/app_namespaces) in the Ably dashboard, add a new rule and enable the **Push notifications** option.
```

Author: done

6. Upload your Firebase service account JSON file in **Setting up Firebase Cloud Messaging** section.
7. In the [Apple Developer portal](https://developer.apple.com), go to **Certificates, Identifiers & Profiles** → **Keys**.
8. Add a new key and check **Apple Push Notifications service (APNs)**, click **Register**.
9. Download the `.p8` file — you can only download it once. Note your **Key ID** and **Team ID**.
10. In the Firebase Console, go to **Project configuration** → **Cloud Messaging** → **Apple App Setup** → **APNS authentication key** to upload your `.p8` file.
11. Register an iOS app in your Firebase project using your bundle identifier. Download `GoogleService-Info.plist` and add it to your Xcode project's root target.

### Create a React Native project <a id="create-project"/>

Create a new React Native project and install the required dependencies:

<Code>
```shell
npx react-native@latest init PushTutorial
cd PushTutorial
npm install ably @react-native-firebase/app @react-native-firebase/messaging @notifee/react-native
```
</Code>

#### Configure Android project <a id="configure-android"/>

Apply the Google Services plugin in `android/build.gradle`:

> ```kotlin

Reviewer: Missing `<Code>` tags on a view code blocks here.

Author: fixed

Open `ios/PushTutorial.xcworkspace` in Xcode and add the `Push Notifications` capability: select your target, go to **Signing & Capabilities**, and click **+ Capability**.

Add `use_modular_headers!` to `ios/Podfile` after `prepare_react_native_project!`:

```text
prepare_react_native_project!
use_modular_headers!
```

This is required for Firebase Swift pods (`FirebaseCoreInternal`, `GoogleUtilities`) to be integrated as static libraries. Then install the native pods:

<Code>
```shell
cd ios && pod install && cd ..
```
</Code>

Then add `FirebaseApp.configure()` to `AppDelegate.swift` before React Native starts:

```swift
import Firebase

func application(
  _ application: UIApplication,
  didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]? = nil
) -> Bool {
  FirebaseApp.configure()
  // ... rest of existing setup
}
```

Add all further code to `App.tsx`.

## Step 1: Set up Ably <a id="step-1"/>

Replace the contents of `App.tsx` with the following to initialize the Ably Pub/Sub client, wrap the app in `AblyProvider` and `ChannelProvider`, and subscribe to a channel for incoming messages using the `useChannel` hook:

<Code>

> ```javascript

Reviewer 2: shouldn't all these code blocks be react, since this is a react native guide?

Author: fixed

import React, {useRef, useState} from 'react';
import {
  Platform,
  ScrollView,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';
import {SafeAreaView} from 'react-native-safe-area-context';
import * as Ably from 'ably';
import {AblyProvider, ChannelProvider, useChannel, useAbly} from 'ably/react';
import messaging from '@react-native-firebase/messaging';
import notifee, {AuthorizationStatus} from '@notifee/react-native';

> const CHANNEL_NAME = 'exampleChannel1';

Reviewer: Can we set this to something like `my-first-push-channel` everywhere for consistency with the other getting started guides?

Author: done

  );
}

// Save the device registration with Ably using the FCM token
async function saveDeviceRegistration(token: string) {
  await client.push.admin.deviceRegistrations.save({
    id: deviceId,
    clientId: 'push-tutorial-client',
    platform: Platform.OS === 'ios' ? 'ios' : 'android',
    formFactor: 'phone',
    push: {
      recipient: {
        transportType: 'fcm', // FCM on both platforms — messaging().getToken() returns an FCM token even on iOS
        registrationToken: token,
      },
    },
  });
}

async function activatePush() {
  try {
    showStatus('Activating push notifications...');
    const granted = await requestPermission();
    if (!granted) {
      showStatus('Notification permission denied.');
      return;
    }

    await messaging().registerDeviceForRemoteMessages(); // Required to receive push notifications on iOS, no-op on Android
    const fcmToken = await messaging().getToken();
    await saveDeviceRegistration(fcmToken);

    tokenRefreshUnsubscribeRef.current?.();
    tokenRefreshUnsubscribeRef.current = messaging().onTokenRefresh(async newToken => {
      try {
        await saveDeviceRegistration(newToken);

>       } catch (error: any) {

Reviewer 2: I believe the call here should return an errorInfo, perhaps worth making that clear in a cast?

Author: done

</Aside>

## Step 3: Test admin push notifications <a id="step-3"/>

The FCM SDK handles background push notifications automatically and displays them as system notifications. For foreground handling, use `@notifee/react-native` to display notifications while the app is open.

Add the following inside `PushScreen`:

<Code>
```javascript
useEffect(() => {
  // Create a default Android notification channel
  if (Platform.OS === 'android') {
    notifee.createChannel({id: 'default', name: 'Default Channel'});
  }

  // Handle foreground push messages
  const unsubscribe = messaging().onMessage(async remoteMessage => {
    const title = remoteMessage.notification?.title ?? 'Push Notification';
    const body = remoteMessage.notification?.body ?? '';
    log(`Push received: ${title} — ${body}`);
    await notifee.displayNotification({
      title,
      body,
      android: {channelId: 'default'},
    });
  });

  return () => {
    unsubscribe();
  };
}, []);
```
</Code>

Use the Ably CLI to send a test push notification to your client ID:

<Code>
```shell

> ably push publish --client-id push-tutorial-client \

Reviewer 2: What version of the cli are you using here? 👀

Author: I'm using `client-id` flag, which is described here - https://ably.com/docs/cli/push/publish#client-id
Or what do you mean?

        {log.map((entry, i) => (
          <Text key={i} style={styles.logEntry}>{entry}</Text>
        ))}
      </ScrollView>
      <TouchableOpacity style={[styles.btn, styles.btnGray]} onPress={() => setLog([])}>
        <Text style={styles.btnText}>Clear Log</Text>
      </TouchableOpacity>
    </View>
  </SafeAreaView>
);
```
</Code>

Add the button styles to the `StyleSheet.create` call:

<Code>
```javascript
buttons: {gap: 8, marginBottom: 12},
btn: {padding: 14, borderRadius: 6, alignItems: 'center'},
btnText: {color: '#fff', fontWeight: '600'},
btnGreen: {backgroundColor: '#28a745'},
btnRed: {backgroundColor: '#dc3545'},
btnPurple: {backgroundColor: '#6f42c1'},
btnOrange: {backgroundColor: '#fd7e14'},
btnGray: {backgroundColor: '#6c757d', marginTop: 8},

```
</Code>

Build and run your app on a physical device:

<Code>
```shell
# Android
npx react-native run-android

# iOS
npx react-native run-ios --device
```
</Code>

You can also open each platform project in their respective IDEs and run from there.

Tap **Activate Push** and wait until the status message displays your device ID. Try sending a test push notification using the Ably CLI commands shown in Step 3.

> ### Send push via channel <a id="step-5-send-channel"/>

Reviewer: Should this be a step 6 in its own right rather than a sub section of building your UI? 

Then we could have sub-sections for publishing via a channel and publishing directly to a device (or client). This might make more sense than having the publish directly to a client part under "test admin notifications" - which we should definitely rename as well.

Author: The point of "Step 3: Test admin push notifications" was so user could see pushes at work rather early in the guide. WDYT @m-hulbert Also what would be the alternative name? (because user should understand that this is admin push, not real app use case)

> ### Send push via channel <a id="step-5-send-channel"/>

Reviewer: ```suggestion
## Step 6: Publish a push notification via a channel <a id="step-5-send-channel"/>
```

We should probably use 'publish' throughout since that's what the existing docs use.

Author: done

### Create a React Native project <a id="create-project"/>

Create a new React Native project and install the required dependencies:

<Code>
```shell
npx react-native@latest init PushTutorial
cd PushTutorial
npm install ably @react-native-firebase/app @react-native-firebase/messaging @notifee/react-native
```
</Code>

#### Configure Android project <a id="configure-android"/>

Apply the Google Services plugin in `android/build.gradle`:

```kotlin
// android/build.gradle
buildscript {
    dependencies {
        classpath('com.google.gms:google-services:4.4.2')
    }
}
```

Then apply it in `android/app/build.gradle`:

```kotlin
// android/app/build.gradle
apply plugin: 'com.google.gms.google-services'
```

Also declare the `POST_NOTIFICATIONS` permission in `android/app/src/main/AndroidManifest.xml`:

```xml
<manifest xmlns:android="http://schemas.android.com/apk/res/android">
    <uses-permission android:name="android.permission.INTERNET" />
    <uses-permission android:name="android.permission.POST_NOTIFICATIONS" />
    ...
</manifest>
```

<Aside data-type='note'>

> At the time of writing React Native Gradle plugin is not compatible with Gradle 9. If you see a `JvmVendorSpec.IBM_SEMERU` error during the Android build, update `android/gradle/wrapper/gradle-wrapper.properties` to use Gradle 8.13: `distributionUrl=https\://services.gradle.org/distributions/gradle-8.13-bin.zip`.

Reviewer: We should avoid things like 'at time of writing' etc. Can we just state we should use 8.13 instead if it's still not supported in 9?

Author: I'm now thinking that it's an insignificant implementation detail which can be easily fixed by android developer familiar with android studio enough to dare implement pushes. I'm removing this.

</Aside>

#### Configure iOS project <a id="configure-ios"/>

Open `ios/PushTutorial.xcworkspace` in Xcode and add the `Push Notifications` capability: select your target, go to **Signing & Capabilities**, and click **+ Capability**.

Add `use_modular_headers!` to `ios/Podfile` after `prepare_react_native_project!`:

> ```ruby

Reviewer: I don't think this should be ruby?

Author: Changed to `text`.

prepare_react_native_project!
use_modular_headers!
```

This is required for Firebase Swift pods (`FirebaseCoreInternal`, `GoogleUtilities`) to be integrated as static libraries. Then install the native pods:

<Code>
```shell
cd ios && pod install && cd ..
```
</Code>

Then add `FirebaseApp.configure()` to `AppDelegate.swift` before React Native starts:

```swift
import Firebase

func application(
  _ application: UIApplication,
  didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]? = nil
) -> Bool {
  FirebaseApp.configure()
  // ... rest of existing setup
}
```

Add all further code to `App.tsx`.

## Step 1: Set up Ably <a id="step-1"/>

> Replace the contents of `App.tsx` with the following to initialize the Ably Realtime client, wrap the app in `AblyProvider` and `ChannelProvider`, and subscribe to a channel for incoming messages using the `useChannel` hook:

Reviewer: Should we keep this to the Pub/Sub rather than a 'realtime client' so people know what it is?

Author: fixed

      return;
    }

    await messaging().registerDeviceForRemoteMessages(); // Required to receive push notifications on iOS, no-op on Android
    const fcmToken = await messaging().getToken();
    await saveDeviceRegistration(fcmToken);

    tokenRefreshUnsubscribeRef.current?.();
    tokenRefreshUnsubscribeRef.current = messaging().onTokenRefresh(async newToken => {
      try {
        await saveDeviceRegistration(newToken);
      } catch (error: any) {
        console.error('Failed to update FCM token:', error.message);
      }
    });

    showStatus(`Push activated. Device ID: ${deviceId}`);
    log(`Push activated. Device ID: ${deviceId}`);
  } catch (error: any) {
    showStatus(`Failed to activate push: ${error.message}`);
  }
}

async function deactivatePush() {
  try {
    showStatus('Deactivating push notifications...');
    await client.push.admin.deviceRegistrations.remove(deviceId);
    tokenRefreshUnsubscribeRef.current?.();
    tokenRefreshUnsubscribeRef.current = null;
    showStatus('Push notifications deactivated.');
  } catch (error: any) {
    showStatus(`Failed to deactivate push: ${error.message}`);
  }
}
```
</Code>

> The `transportType` is set to `fcm` on both platforms because `messaging().getToken()` always returns an FCM registration token, even on iOS. Firebase exchanges the APNs device token for an FCM token internally, so Ably communicates with Firebase rather than APNs directly (instead Firebase sends push notifications to iOS devices via APNs).

Reviewer: Do we need to explain this? Would someone know this if they're React Native devs or familiar with Push? Maybe even a short code comment for iOS would be sufficient?

Author: Moved to code comment (I'm not sure what you've meant by "for iOS" tho, it's cross-platform js code snippets).

`activatePush()` does the following:

1. Requests notification permission from the user.
2. Obtains the FCM registration token from Firebase.
3. Registers the device with Ably's push notification service using the token.

After successful activation, `deviceId` contains the unique device ID assigned to this installation.

> <Aside data-type='note'>

Reviewer: Do we need all this exposition? Three notes in a row suggests this is way too complex. Do we need to explain which methods are used and why?

Author: Moved two of them in a short code comment.

The `deviceId` used here is randomly generated on each app start for simplicity. In a production app, persist the device ID (for example using `AsyncStorage`) so that the same device registration is reused across restarts. Without persistence, each app restart creates a new device registration with Ably, orphaning the previous one.
</Aside>

<Aside data-type='note'>
On Android, use `notifee.requestPermission()` instead of React Native's `PermissionsAndroid` because it handles the `POST_NOTIFICATIONS` permission consistently across Android API levels. On iOS, `messaging().requestPermission()` shows the system notification permission dialog; Firebase Messaging calls the native iOS permission APIs internally.
</Aside>

<Aside data-type='note'>
In React Native `realtime.push.activate()` is not used because it's a JS SDK and it registers browser for push notifications, which is not what you want in a React Native app. Instead, use `realtime.push.admin.deviceRegistrations` API to register the device.
</Aside>

## Step 3: Subscribe and test push notifications <a id="step-3"/>

> The FCM SDK handles background push notifications automatically and displays them as system notifications. For foreground handling, use `@notifee/react-native` to display notifications while the app is open.

Reviewer: Should we choose one approach or the other? Or explain how to use both.

Author: I've meant that if a reader wants to display pushes inside their app (foreground), they should use notifee (code example already shows how). If app in the background though, it will be handled by Android itself and displayed in the Notification Center.

    const body = remoteMessage.notification?.body ?? '';
    log(`Push received: ${title} — ${body}`);
    await notifee.displayNotification({
      title,
      body,
      android: {channelId: 'default'},
    });
  });

  return () => {
    unsubscribe();
  };
}, []);
```
</Code>

To subscribe your device to a channel so it can receive channel-based push notifications, add the following functions inside `PushScreen`, after the functions from Step 2:

<Code>
```javascript
async function subscribeToChannel() {
  try {
    await client.push.admin.channelSubscriptions.save({
      deviceId,
      channel: CHANNEL_NAME,
    });
    showStatus(`Subscribed to push on channel: ${CHANNEL_NAME}`);
  } catch (error: any) {
    showStatus(`Failed to subscribe: ${error.message}`);
  }
}

async function unsubscribeFromChannel() {
  try {
    await client.push.admin.channelSubscriptions.remove({
      deviceId,
      channel: CHANNEL_NAME,
    });
    showStatus(`Unsubscribed from push on channel: ${CHANNEL_NAME}`);
  } catch (error: any) {
    showStatus(`Failed to unsubscribe: ${error.message}`);
  }
}
```
</Code>

<Aside data-type='note'>

> Sending push notifications using `deviceId` or `clientId` requires the `push-admin` capability for your API key. Use this method for testing purposes. In production, send push notifications from your backend server by publishing messages with a `push` `extras` field to a channel.

Reviewer: Could the capabilities be covered in the pre-reqs rather than mid-guide? Which method are we then referring to here?

Author: Moved to Prerequisites.

        ))}
      </ScrollView>
      <TouchableOpacity style={[styles.btn, styles.btnGray]} onPress={() => setLog([])}>
        <Text style={styles.btnText}>Clear Log</Text>
      </TouchableOpacity>
    </View>
  </SafeAreaView>
);
```
</Code>

Add the button styles to the `StyleSheet.create` call:

<Code>
```javascript
buttons: {gap: 8, marginBottom: 12},
btn: {padding: 14, borderRadius: 6, alignItems: 'center'},
btnText: {color: '#fff', fontWeight: '600'},
btnGreen: {backgroundColor: '#28a745'},
btnRed: {backgroundColor: '#dc3545'},
btnPurple: {backgroundColor: '#6f42c1'},
btnOrange: {backgroundColor: '#fd7e14'},
btnGray: {backgroundColor: '#6c757d', marginTop: 8},

```
</Code>

Build and run your app on a physical device:

<Code>
```shell
# Android
npx react-native run-android

# iOS
npx react-native run-ios --device
```
</Code>

You can also open each platform project in their respective IDEs and run from there.

Tap **Activate Push** and wait until the status message displays your device ID. Try sending a test push notification using the Ably CLI commands shown in Step 3.

### Send push via channel <a id="step-4-send-channel"/>

> To test push notifications via channel, tap **Subscribe to Channel** in the app and then send a push notification to the channel using the Ably CLI:

Reviewer: There's a lot of 'testing' here and do we need both options? I think I agree with Steven that we should either use the dashboard (or more likely the CLI) for pushing the notifications as you wouldn't tend to be sending via the app in reality.

Author: Not sure I follow, do you mean to remove code snippet for js and only leave CLI? (js one was Steven suggestion.)

Reviewer: I was referring to showing how to publish via code and also using the CLI. Since this is a simple getting started guide I think we should show 1 or the other, rather than all the options. I'd suggest using the CLI for simplicity (and then you could note or link to the code equivalent in the existing docs).

Author: done

### Create a React Native project <a id="create-project"/>

Create a new React Native project and install the required dependencies:

<Code>
```shell
npx react-native@latest init PushTutorial
cd PushTutorial
npm install ably @react-native-firebase/app @react-native-firebase/messaging @notifee/react-native
```
</Code>

#### Configure Android project <a id="configure-android"/>

Apply the Google Services plugin in `android/build.gradle`:

<Code>
```kotlin
// android/build.gradle
buildscript {
    dependencies {
        classpath('com.google.gms:google-services:4.4.2')
    }
}
```
</Code>

Then apply it in `android/app/build.gradle`:

<Code>
```kotlin
// android/app/build.gradle
apply plugin: 'com.google.gms.google-services'
```
</Code>

Also declare the `POST_NOTIFICATIONS` permission in `android/app/src/main/AndroidManifest.xml`:

<Code>
```xml
<manifest xmlns:android="http://schemas.android.com/apk/res/android">
    <uses-permission android:name="android.permission.INTERNET" />
    <uses-permission android:name="android.permission.POST_NOTIFICATIONS" />
    ...
</manifest>
```
</Code>

> <Aside>

Reviewer 3: This has no data-type attribute. All asides should. It could just be <Aside data-type='note'>

Author: Fixed, added `data-type='note'`.

> At the time of writing React Native 0.84's Gradle plugin is not compatible with Gradle 9. If you see a `JvmVendorSpec.IBM_SEMERU` error during the Android build, update `android/gradle/wrapper/gradle-wrapper.properties` to use Gradle 8.13: `distributionUrl=https\://services.gradle.org/distributions/gradle-8.13-bin.zip`.

Author: I just removed version number.
